curvefs/src/metaserver/recycle_cleaner.cpp

```diff
@@ -0,0 +1,382 @@
+        parent != 0) {
+        auto parents = inode.mutable_parent();
+        for (auto iter = parents->begin(); iter != parents->end(); iter++) {
+            if (*iter == parent) {
+                parents->erase(iter);
+                break;
+            }
+        }
+    }
+
+    *dirtyAttr.mutable_parent() = inode.parent();
+    ret = metaClient_->UpdateInodeAttr(fsId, inodeid, dirtyAttr);
+    if (ret != MetaStatusCode::OK) {
+        LOG(WARNING) << "UpdateInodeAttr failed, ret = "
+                     << MetaStatusCode_Name(ret)
+                     << ", inodeid = " << inode.inodeid();
+        return false;
+    }
+
+    // when use recycle, should disable SumInDir
+
+    return true;
+}
+
+// delete file under dir, delete dir
+bool RecycleCleaner::DeleteDirRecursive(const Dentry& dentry) {
+    uint32_t fsId = partition_->GetFsId();
+    uint64_t inodeid = dentry.inodeid();
+    std::string last = "";
+    bool onlyDir = false;
+
+    while (true) {
+        // 1. list dir
+        std::list<Dentry> dentryList;
+        auto ret = metaClient_->ListDentry(fsId, inodeid, last, limit_, onlyDir,
+                                           &dentryList);
+        if (ret != MetaStatusCode::OK) {
+            LOG(WARNING) << "DeleteDirRecursive list dentry fail, ret = "
+                         << MetaStatusCode_Name(ret)
+                         << ", dentry: " << dentry.ShortDebugString();
+            return false;
+        }
+        // 2. delete file under dir
+        for (const auto& it : dentryList) {
+            if (isStop_ || !copysetNode_->IsLeaderTerm()) {
+                LOG(WARNING) << "recycle cleaner stop or not leader, isStop = "
+                             << isStop_
+                             << ", isLeader = " << copysetNode_->IsLeaderTerm();
+                return false;
+            }
+
+            FsFileType type;
+            if (it.has_type()) {
+                type = it.type();
+            } else {
+                InodeAttr attr;
+                ret = metaClient_->GetInodeAttr(fsId, it.inodeid(), &attr);
+                if (ret != MetaStatusCode::OK) {
+                    LOG(WARNING)
+                        << "DeleteDirRecursive get inode attr fail, ret = "
+                        << MetaStatusCode_Name(ret) << ", fsId = " << fsId
+                        << ", inodeid = " << it.inodeid();
+                    return false;
+                }
+                type = attr.type();
+            }
+
+            if (type == FsFileType::TYPE_DIRECTORY) {
+                // if type is diretory, delete recursive
+                if (!DeleteDirRecursive(it)) {
+                    LOG(WARNING) << "DeleteDirRecursive delete sub dentry fail,"
+                                 << " dentry: " << it.ShortDebugString();
+                    return false;
+                }
+            } else {
+                // if type is not directory, delete direct
+                if (!DeleteNode(it)) {
```

Comment:
Will there be any problems if the leader is switched during the deletion process?

Reply:
This will result in orphaned inodes.
The normal file deletion process also has this problem.